File: apps/windows/tests/Polyglance.Platform.Tests/PinSessionTests.cs
using System;
using System.IO;
using System.Linq;
using System.Windows.Media;
using System.Windows.Media.Imaging;
using Polyglance.Platform.Pin;
using Xunit;

namespace Polyglance.Platform.Tests;

public sealed class PinSessionTests : IDisposable
{
    [Fact]
    public void SessionMemoryBudgetUsesDecodedImageDimensions()
    {
        var store = new PinArchiveStore(_directory);
        Assert.NotNull(store.Append(Image(), PinArchiveSource.Clipboard));
        string path = Path.Combine(_directory, "index.json");
        var json = System.Text.Json.Nodes.JsonNode.Parse(File.ReadAllText(path))!;
        json["items"]![0]!["pixelWidth"] = int.MaxValue;
        json["items"]![0]!["pixelHeight"] = int.MaxValue;
        File.WriteAllText(path, json.ToJsonString());
        var item = Assert.Single(store.List());
        Assert.Equal(8, item.PixelWidth);
        Assert.Equal(8, item.PixelHeight);
    }
    [Fact]
    public void MissingIndexDoesNotLetDestroyOrClearLeaveOrphanImages()
    {
        var store = new PinArchiveStore(_directory);
        var item = store.Append(Image(), PinArchiveSource.Clipboard)!;
        File.Delete(Path.Combine(_directory, "index.json"));
        Assert.True(store.Delete(item.Id).IsSuccess);
        Assert.Empty(store.List());
        Assert.NotNull(store.Append(Image(), PinArchiveSource.Clipboard));
        File.Delete(Path.Combine(_directory, "index.json"));
        Assert.True(store.DeleteAll().IsSuccess);
        Assert.Empty(store.List());
    }
    private readonly string _directory = Path.Combine(Path.GetTempPath(), "PolyglanceSessionTest-" + Guid.NewGuid());
    private static BitmapSource Image()
    {
        var image = BitmapSource.Create(8, 8, 96, 96, PixelFormats.Bgra32, null, new byte[8 * 8 * 4], 32);
        image.Freeze();
        return image;
    }

    [Fact]
    public void SessionRoundTripAndDestroyRemovesOriginalText()
    {
        var store = new PinArchiveStore(_directory);
        var item = store.Append(Image(), PinArchiveSource.Clipboard)!;
        var state = new PinSessionRecord { ArchiveId = item.Id, Text = "original text", X = -500, Y = 90, Width = 300, Height = 200, Opacity = .6 };
        Assert.True(store.SaveSession(state));
        var loaded = new PinArchiveStore(_directory).LoadSessions().Single();
        Assert.Equal(state.Text, loaded.Text);
        Assert.Equal(state.X, loaded.X);
        Assert.Equal(state.Opacity, loaded.Opacity);
        Assert.True(store.Delete(item.Id).IsSuccess);
        Assert.Empty(store.LoadSessions());
        Assert.False(store.SaveSession(state));
        Assert.DoesNotContain("original text", File.ReadAllText(Path.Combine(_directory, "sessions.json")));
    }

    [Fact]
    public void ActiveImageIsNotEvictedByNewHistory()
    {
        var store = new PinArchiveStore(_directory, maxCount: 2);
        var active = store.Append(Image(), PinArchiveSource.Clipboard)!;
        Assert.True(store.SaveSession(new PinSessionRecord { ArchiveId = active.Id, Width = 100, Height = 100 }));
        var old = store.Append(Image(), PinArchiveSource.Screenshot)!;
        Assert.NotNull(store.Append(Image(), PinArchiveSource.Screenshot));
        Assert.NotNull(store.LoadImage(active.Id));
        Assert.Null(store.LoadImage(old.Id));
    }

    public void Dispose() { if (Directory.Exists(_directory)) Directory.Delete(_directory, true); }

    [Fact]
    public void ClosedHistoryIsBoundedAndClearRemovesText()
    {
        var store = new PinArchiveStore(_directory);
        var item = store.Append(Image(), PinArchiveSource.Clipboard)!;
        for (int number = 0; number < 25; number++)
            Assert.True(store.SaveSession(new PinSessionRecord { ArchiveId = item.Id, Text = "text-" + number, Status = PinSessionStatus.Closed }));
        Assert.Equal(20, store.LoadSessions().Count(s => s.Status == PinSessionStatus.Closed));
        Assert.Equal("text-24", store.LoadSessions().Last().Text);
        Assert.True(store.DeleteAll().IsSuccess);
        Assert.Empty(store.LoadSessions());
        Assert.DoesNotContain("text-", File.ReadAllText(Path.Combine(_directory, "sessions.json")));
    }

    [Fact]
    public void FullActiveArchiveRejectsNewImageWithoutEviction()
    {
        var store = new PinArchiveStore(_directory, maxCount: 1);
        var active = store.Append(Image(), PinArchiveSource.Clipboard)!;
        Assert.True(store.SaveSession(new PinSessionRecord { ArchiveId = active.Id }));
        Assert.Null(store.Append(Image(), PinArchiveSource.Screenshot));
        Assert.Equal(active.Id, Assert.Single(store.List()).Id);
    }

    [Fact]
    public void InvalidStateIsRejectedAndReplacementCannotRecreateDeletedImage()
    {
        var store = new PinArchiveStore(_directory);
        var item = store.Append(Image(), PinArchiveSource.Clipboard)!;
        Assert.False(store.SaveSession(new PinSessionRecord { ArchiveId = item.Id, Opacity = double.NaN }));
        Assert.True(store.ReplaceImage(Image(), item.Id));
        Assert.True(store.Delete(item.Id).IsSuccess);
        Assert.False(store.ReplaceImage(Image(), item.Id));
        Assert.True(store.Delete(item.Id).IsSuccess);
    }

    [Fact]
    public void CorruptIndexRecoveryKeepsSessionIdentity()
    {
        var store = new PinArchiveStore(_directory);
        var item = store.Append(Image(), PinArchiveSource.Clipboard)!;
        Assert.True(store.SaveSession(new PinSessionRecord { ArchiveId = item.Id, Text = "recover me" }));
        File.WriteAllText(Path.Combine(_directory, "index.json"), "damaged index");
        Assert.Equal(item.Id, Assert.Single(store.List()).Id);
        Assert.Equal("recover me", Assert.Single(store.LoadSessions()).Text);
    }

    [Fact]
    public void CorruptSessionMetadataIsNotOverwritten()
    {
        var store = new PinArchiveStore(_directory);
        var item = store.Append(Image(), PinArchiveSource.Clipboard)!;
        string path = Path.Combine(_directory, "sessions.json");
        File.WriteAllText(path, "damaged metadata");
        Assert.False(store.SaveSession(new PinSessionRecord { ArchiveId = item.Id }));
        Assert.Equal("damaged metadata", File.ReadAllText(path));
    }
}

File: apps/macos/Tests/PolyglanceTests/PinSessionTests.swift
import AppKit
import XCTest
@testable import Polyglance

final class PinSessionTests: XCTestCase {
    func testSessionMemoryBudgetUsesDecodedImageDimensions() throws {
        let dir = FileManager.default.temporaryDirectory.appendingPathComponent(UUID().uuidString)
        defer { try? FileManager.default.removeItem(at: dir) }
        let store = PinArchiveStore(directoryURL: dir)
        _ = try XCTUnwrap(store.append(image: image(), source: .clipboard))
        let path = dir.appendingPathComponent("index.json")
        var manifest = try XCTUnwrap(JSONSerialization.jsonObject(with: Data(contentsOf: path)) as? [String: Any])
        var items = try XCTUnwrap(manifest["items"] as? [[String: Any]])
        items[0]["pixelWidth"] = Int.max
        items[0]["pixelHeight"] = Int.max
        manifest["items"] = items
        try JSONSerialization.data(withJSONObject: manifest).write(to: path)
        let loaded = try XCTUnwrap(store.list().first)
        XCTAssertEqual(loaded.pixelWidth, 8)
        XCTAssertEqual(loaded.pixelHeight, 8)
    }
    func testMissingIndexDoesNotLetDestroyOrClearLeaveOrphanImages() throws {
        let dir = FileManager.default.temporaryDirectory.appendingPathComponent(UUID().uuidString)
        defer { try? FileManager.default.removeItem(at: dir) }
        let store = PinArchiveStore(directoryURL: dir)
        let item = try XCTUnwrap(store.append(image: image(), source: .clipboard))
        try FileManager.default.removeItem(at: dir.appendingPathComponent("index.json"))
        XCTAssertTrue(store.delete(id: item.id).isSuccess)
        XCTAssertTrue(store.list().isEmpty)
        _ = try XCTUnwrap(store.append(image: image(), source: .clipboard))
        try FileManager.default.removeItem(at: dir.appendingPathComponent("index.json"))
        XCTAssertNil(store.deleteAll().errorMessage)
        XCTAssertTrue(store.list().isEmpty)
    }
    private func image() -> NSImage {
        NSImage(size: NSSize(width: 8, height: 8), flipped: false) { rect in
            NSColor.white.setFill(); rect.fill(); return true
        }
    }

    func testClosedHistoryIsBoundedAndClearRemovesText() throws {
        let dir = FileManager.default.temporaryDirectory.appendingPathComponent(UUID().uuidString)
        defer { try? FileManager.default.removeItem(at: dir) }
        let store = PinArchiveStore(directoryURL: dir)
        let item = try XCTUnwrap(store.append(image: image(), source: .clipboard))
        for number in 0..<25 {
            var record = PinSessionRecord(archiveID: item.id, text: "text-\(number)", frame: CGRect(x: number, y: 0, width: 200, height: 100))
            record.status = .closed
            XCTAssertTrue(store.saveSession(record))
        }
        XCTAssertEqual(store.loadSessions().filter { $0.status == .closed }.count, 20)
        XCTAssertEqual(store.loadSessions().last?.text, "text-24")
        XCTAssertNil(store.deleteAll().errorMessage)
        XCTAssertTrue(store.loadSessions().isEmpty)
        XCTAssertFalse(try String(contentsOf: dir.appendingPathComponent("sessions.json"), encoding: .utf8).contains("text-"))
    }

    func testFullActiveArchiveRejectsNewRecordAndDoesNotEvictActive() throws {
        let dir = FileManager.default.temporaryDirectory.appendingPathComponent(UUID().uuidString)
        defer { try? FileManager.default.removeItem(at: dir) }
        let store = PinArchiveStore(directoryURL: dir, maximumCount: 1)
        let active = try XCTUnwrap(store.append(image: image(), source: .clipboard))
        XCTAssertTrue(store.saveSession(PinSessionRecord(archiveID: active.id, frame: CGRect(x: 0, y: 0, width: 100, height: 100))))
        XCTAssertNil(store.append(image: image(), source: .screenshot))
        XCTAssertEqual(store.list().map(\.id), [active.id])
    }

    func testReplacementCannotRecreateDeletedImageAndInvalidStateIsRejected() throws {
        let dir = FileManager.default.temporaryDirectory.appendingPathComponent(UUID().uuidString)
        defer { try? FileManager.default.removeItem(at: dir) }
        let store = PinArchiveStore(directoryURL: dir)
        let item = try XCTUnwrap(store.append(image: image(), source: .clipboard))
        var state = PinSessionRecord(archiveID: item.id, frame: CGRect(x: 0, y: 0, width: 100, height: 100))
        state.opacity = .nan
        XCTAssertFalse(store.saveSession(state))
        XCTAssertTrue(store.replaceImage(image(), id: item.id))
        XCTAssertTrue(store.delete(id: item.id).isSuccess)
        XCTAssertTrue(store.wasDeleted(item.id))
        XCTAssertFalse(store.replaceImage(image(), id: item.id))
        XCTAssertTrue(store.delete(id: item.id).isSuccess)
        let specialized = try XCTUnwrap(store.append(image: image(), source: .ocr))
        XCTAssertNil(store.deleteAll().errorMessage)
        XCTAssertTrue(store.wasDeleted(specialized.id), "Clear also invalidates in-memory specialized OCR snapshots without session metadata")
    }

    func testCorruptSessionMetadataIsNotOverwritten() throws {
        let dir = FileManager.default.temporaryDirectory.appendingPathComponent(UUID().uuidString)
        defer { try? FileManager.default.removeItem(at: dir) }
        let store = PinArchiveStore(directoryURL: dir)
        let item = try XCTUnwrap(store.append(image: image(), source: .clipboard))
        let url = dir.appendingPathComponent("sessions.json")
        try Data("damaged metadata".utf8).write(to: url)
        XCTAssertFalse(store.saveSession(PinSessionRecord(archiveID: item.id, frame: CGRect(x: 0, y: 0, width: 100, height: 100))))
        XCTAssertEqual(try String(contentsOf: url, encoding: .utf8), "damaged metadata")
    }
    func testSessionRoundTripAndDestroyRemovesOriginalText() throws {
        let dir = FileManager.default.temporaryDirectory.appendingPathComponent(UUID().uuidString)
        defer { try? FileManager.default.removeItem(at: dir) }
        let store = PinArchiveStore(directoryURL: dir)
        let image = NSImage(size: NSSize(width: 20, height: 20), flipped: false) { rect in
            NSColor.white.setFill(); rect.fill(); return true
        }
        let item = try XCTUnwrap(store.append(image: image, source: .clipboard))
        var state = PinSessionRecord(archiveID: item.id, text: "原文\nselect me", frame: CGRect(x: -500, y: 90, width: 300, height: 200))
        state.opacity = 0.6
        XCTAssertTrue(store.saveSession(state))
        XCTAssertEqual(PinArchiveStore(directoryURL: dir).loadSessions().first, state)
        state.status = .closed
        XCTAssertTrue(store.saveSession(state))
        XCTAssertEqual(store.loadSessions().first?.status, .closed)
        XCTAssertTrue(store.delete(id: item.id).isSuccess)
        XCTAssertTrue(store.loadSessions().isEmpty)
        XCTAssertFalse(store.saveSession(state), "An old queued UI update must not recreate deleted history")
        let metadata = try String(contentsOf: dir.appendingPathComponent("sessions.json"), encoding: .utf8)
        XCTAssertFalse(metadata.contains("select me"))
    }

    func testActiveImageIsNotEvictedByNewHistory() throws {
        let dir = FileManager.default.temporaryDirectory.appendingPathComponent(UUID().uuidString)
        defer { try? FileManager.default.removeItem(at: dir) }
        let store = PinArchiveStore(directoryURL: dir, maximumCount: 2)
        let image = NSImage(size: NSSize(width: 8, height: 8), flipped: false) { rect in
            NSColor.white.setFill(); rect.fill(); return true
        }
        let active = try XCTUnwrap(store.append(image: image, source: .clipboard))
        XCTAssertTrue(store.saveSession(PinSessionRecord(archiveID: active.id, frame: CGRect(x: 0, y: 0, width: 100, height: 100))))
        let old = try XCTUnwrap(store.append(image: image, source: .screenshot))
        _ = try XCTUnwrap(store.append(image: image, source: .screenshot))
        XCTAssertNotNil(store.loadImage(id: active.id))
        XCTAssertNil(store.loadImage(id: old.id))
    }
}
